app/test/testmerginapi.cpp

```diff
@@ -3116,3 +3116,112 @@ void TestMerginApi::testDownloadWithNetworkErrorRecovery()
+    QByteArray data = "this is not valid JSON";
+    file.write( data );
+    file.close();
+  }
+  config = MerginConfig::fromFile( tempFilePath );
+  QVERIFY( !config.isValid );
+  QFile::remove( tempFilePath );
+
+  // 3 => empty file
+  tempFilePath = QDir::tempPath() + "/test_empty_config.json";
+  {
+    QFile file( tempFilePath );
+    QVERIFY( file.open( QIODevice::WriteOnly ) );
+    file.write( "" );
+    file.close();
+  }
+  config = MerginConfig::fromFile( tempFilePath );
+  QVERIFY( !config.isValid );
+  QFile::remove( tempFilePath );
+
+  // 4 => file-not-found
+  tempFilePath = QDir::tempPath() + "/nonexistent_config.json";
+  config = MerginConfig::fromFile( tempFilePath );
+  QVERIFY( !config.isValid );
+}
+
+void TestMerginApi::testHasLocalChangesWithSelectiveSyncEnabled()
+{
+  // temporary project directory
+  QTemporaryDir tempDir;
+  QVERIFY( tempDir.isValid() );
+  QString projectDir = tempDir.path();
+
+  // create a mergin‑config file enabling selective sync,
+  // so that files under "photos" are excluded from the sync
+  QString configPath = projectDir + "/mergin-config.json";
+  {
+    QFile configFile( configPath );
+    QVERIFY( configFile.open( QIODevice::WriteOnly ) );
+    configFile.write( "{\"input-selective-sync\": true, \"input-selective-sync-dir\": \"photos\"}" );
+    configFile.close();
+  }
+
+  // create two server files => one in "photos" folder (excluded), one in root folder (not excluded)
+  MerginFile serverExcluded;
+  serverExcluded.path = "photos/photo.jpg";
+  serverExcluded.checksum = "img_checksum";
+
+  MerginFile serverIncluded;
+  serverIncluded.path = "data.txt";
+  serverIncluded.checksum = "data_checksum";
+
+  QList<MerginFile> oldServerFiles;
+  oldServerFiles.append( serverExcluded );
+  oldServerFiles.append( serverIncluded );
+
+  // retrieve config file we wrote
+  MerginConfig config = MerginConfig::fromFile( configPath );
+  QVERIFY( config.isValid );
+  QCOMPARE( config.selectiveSyncEnabled, true );
+  QCOMPARE( config.selectiveSyncDir, QString( "photos" ) ); // verify selective sync folder
+
+  // first scenario => local files list exactly matches the non‑excluded server file
+  // the excluded file ("photos/photo.jpg") is ignored, and no local changes should be detected
+  QList<MerginFile> localFilesNoChange;
+  localFilesNoChange.append( serverIncluded );
+
+  bool result = mApi->hasLocalChanges( oldServerFiles, localFilesNoChange, projectDir, config );
+  QVERIFY( !result );
+
+  // second scenario => local file list contains a modified version of non‑excluded file
+  // the checksum of "data.txt" is different, and local changes should be detected
+  QList<MerginFile> localFilesChanged;
+  {
+    MerginFile modifiedIncluded = serverIncluded;
+    modifiedIncluded.checksum = "different_checksum";
+    localFilesChanged.append( modifiedIncluded );
+  }
+
+  result = mApi->hasLocalChanges( oldServerFiles, localFilesChanged, projectDir, config );
+  QVERIFY( result );
```

Comment:
It would be good to also test the `MerginApi::hasLocalProjectChanges` method. We need to verify that it reads the correct config file and all works as expected. I suggest to create a new test function for that. We could create a new project, update some files, runs this method and check the result. Mind that we need to cover scenarios like - a) selective sync (dis-)enabled, b) there are (not) some local changes